File: iosApp/iosApp/Presentation/Main/Admin/Dormant/DormantAccountViewModel.swift
import Foundation
import Combine
import Shared

@MainActor
final class DormantAccountViewModel: ObservableObject {
    private let getDormantAccountPageUseCase: GetDormantAccountPageUseCase

    private let updateUserDormantStatusUseCase: UpdateUserDormantStatusUseCase

    @Published private(set) var uiState = DormantAccountUiState()

    let event = PassthroughSubject<DormantAccountEvent, Never>()

    private var pageTask: Task<Void, Never>?

    func onAction(_ action: DormantAccountAction) {
        switch action {
        case .selectFilter(let filter):
            guard uiState.selectedFilter != filter else { return }
            uiState.selectedFilter = filter
            uiState.users = []
            uiState.nextCursor = nil
            uiState.canLoadMore = false
            uiState.confirmTarget = nil
            uiState.infoMessage = nil
            loadPage(cursor: nil, append: false)
        case .loadMore:
            loadMore()
        case .requestDormantChange(let user, let dormant):
            uiState.confirmTarget = DormantChangeRequest(user: user, dormant: dormant)
        case .confirmDormantChange:
            confirmDormantChange()
        case .cancelDormantChange:
            uiState.confirmTarget = nil
        case .dismissInfoMessage:
            uiState.infoMessage = nil
        }
    }

    private func loadMore() {
        guard uiState.canLoadMore, !uiState.isLoadingMore, let cursor = uiState.nextCursor else { return }
        loadPage(cursor: cursor, append: true)
    }

    private func loadPage(cursor: String?, append: Bool) {
        pageTask?.cancel()
        pageTask = Task { @MainActor in
            if append {
                uiState.isLoadingMore = true
                do {
                    try await Task.sleep(nanoseconds: dormantPaginationDelayNanoseconds)
                    if Task.isCancelled { return }
                } catch {
                    return
                }
            } else {
                uiState.isLoading = true
                uiState.infoMessage = nil
            }

            do {
                let result = try await getDormantAccountPageUseCase.invoke(
                    filter: uiState.selectedFilter,
                    cursor: cursor,
                    pageSize: dormantAccountPageSize
                )
                if let success = result as? AppResultSuccess<AnyObject>,
                   let page = success.data as? Shared.PagedResult<User> {
                    let items = page.items as? [User] ?? []
                    let pageItems = uiState.selectedFilter == .dormant
                        ? items.sorted { $0.createdAt > $1.createdAt }
                        : items.sorted { ($0.lastLoginAt ?? "") < ($1.lastLoginAt ?? "") }
                    uiState.users = append ? (uiState.users + pageItems) : pageItems
                    uiState.nextCursor = page.nextCursor
                    uiState.canLoadMore = page.hasNext
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                    uiState.infoMessage = nil
                } else if let failure = result as? AppResultFailure {
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                    uiState.infoMessage = "\(failure.error)"
                } else {
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.isLoadingMore = false
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    private func confirmDormantChange() {
        guard let request = uiState.confirmTarget, !uiState.isUpdating else { return }

        uiState.confirmTarget = nil
        Task { @MainActor in
            uiState.isUpdating = true

            do {
                let result = try await updateUserDormantStatusUseCase.invoke(
                    userId: request.user.id,
                    dormant: request.dormant
                )
                if result is AppResultSuccess<AnyObject> {
                    uiState.users = uiState.users.filter { $0.id != request.user.id }
                    uiState.isUpdating = false
                    uiState.infoMessage = request.dormant
                        ? "\(request.user.nickname) 계정을 휴면 전환했습니다."
                        : "\(request.user.nickname) 계정의 휴면을 해제했습니다."
                } else if let failure = result as? AppResultFailure {
                    uiState.isUpdating = false
                    uiState.infoMessage = "\(failure.error)"
                } else {
                    uiState.isUpdating = false
                }
            } catch {
                uiState.isUpdating = false
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    init(
        getDormantAccountPageUseCase: GetDormantAccountPageUseCase = KoinInitializerKt.resolveGetDormantAccountPageUseCase(),
        updateUserDormantStatusUseCase: UpdateUserDormantStatusUseCase = KoinInitializerKt.resolveUpdateUserDormantStatusUseCase()
    ) {
        self.getDormantAccountPageUseCase = getDormantAccountPageUseCase
        self.updateUserDormantStatusUseCase = updateUserDormantStatusUseCase

        loadPage(cursor: nil, append: false)
    }

    deinit {
        pageTask?.cancel()
    }
}

private let dormantAccountPageSize: Int32 = 15
private let dormantPaginationDelayNanoseconds: UInt64 = 1_000_000_000

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/admin/dormant/DormantAccountViewModel.kt
package com.hhp227.concafe.presentation.main.admin.dormant

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.DormantAccountFilter
import com.hhp227.concafe.domain.usecase.GetDormantAccountPageUseCase
import com.hhp227.concafe.domain.usecase.UpdateUserDormantStatusUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class DormantAccountViewModel(
    private val getDormantAccountPageUseCase: GetDormantAccountPageUseCase,
    private val updateUserDormantStatusUseCase: UpdateUserDormantStatusUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(DormantAccountUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<DormantAccountEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private var pageJob: Job? = null

    fun onAction(action: DormantAccountAction) {
        when (action) {
            is DormantAccountAction.SelectFilter -> {
                if (_uiState.value.selectedFilter != action.filter) {
                    _uiState.update {
                        it.copy(
                            selectedFilter = action.filter,
                            users = emptyList(),
                            nextCursor = null,
                            canLoadMore = false,
                            confirmTarget = null,
                            infoMessage = null
                        )
                    }
                    loadPage(cursor = null, append = false)
                }
            }
            DormantAccountAction.LoadMore -> loadMore()
            is DormantAccountAction.RequestDormantChange -> {
                _uiState.update {
                    it.copy(confirmTarget = DormantChangeRequest(user = action.user, dormant = action.dormant))
                }
            }
            DormantAccountAction.ConfirmDormantChange -> confirmDormantChange()
            DormantAccountAction.CancelDormantChange -> {
                _uiState.update { it.copy(confirmTarget = null) }
            }
            DormantAccountAction.DismissInfoMessage -> {
                _uiState.update { it.copy(infoMessage = null) }
            }
        }
    }

    private fun loadMore() {
        val state = _uiState.value
        val cursor = state.nextCursor
        if (cursor == null || !state.canLoadMore || state.isLoadingMore) return
        loadPage(cursor = cursor, append = true)
    }

    private fun loadPage(cursor: String?, append: Boolean) {
        pageJob?.cancel()
        pageJob = viewModelScope.launch {
            if (append) {
                _uiState.update { it.copy(isLoadingMore = true) }
                delay(PAGINATION_DELAY_MILLIS)
            } else {
                _uiState.update { it.copy(isLoading = true, infoMessage = null) }
            }

            when (val result = getDormantAccountPageUseCase.invoke(
                filter = _uiState.value.selectedFilter,
                cursor = cursor,
                pageSize = DORMANT_ACCOUNT_PAGE_SIZE
            )) {
                is AppResult.Success -> {
                    val pageItems = if (_uiState.value.selectedFilter == DormantAccountFilter.DORMANT) {
                        result.data.items.sortedByDescending { it.createdAt }
                    } else {
                        result.data.items.sortedBy { it.lastLoginAt }
                    }
                    _uiState.update { state ->
                        state.copy(
                            users = if (append) state.users + pageItems else pageItems,
                            nextCursor = result.data.nextCursor,
                            canLoadMore = result.data.hasNext,
                            isLoading = false,
                            isLoadingMore = false,
                            infoMessage = null
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            isLoadingMore = false,
                            infoMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    private fun confirmDormantChange() {
        val request = _uiState.value.confirmTarget

        if (request == null || _uiState.value.isUpdating) {
            return
        }
        viewModelScope.launch {
            _uiState.update { it.copy(isUpdating = true) }

            when (val result = updateUserDormantStatusUseCase.invoke(
                userId = request.user.id,
                dormant = request.dormant
            )) {
                is AppResult.Success -> {
                    _uiState.update { state ->
                        state.copy(
                            users = state.users.filterNot { it.id == request.user.id },
                            isUpdating = false,
                            confirmTarget = null,
                            infoMessage = if (request.dormant) {
                                "${request.user.nickname} 계정을 휴면 전환했습니다."
                            } else {
                                "${request.user.nickname} 계정의 휴면을 해제했습니다."
                            }
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isUpdating = false,
                            confirmTarget = null,
                            infoMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    override fun onCleared() {
        pageJob?.cancel()
        super.onCleared()
    }

    init {
        loadPage(cursor = null, append = false)
    }
}

private const val DORMANT_ACCOUNT_PAGE_SIZE = 15
private const val PAGINATION_DELAY_MILLIS = 1_000L
